### ISSUE TITLE
Unexpected behavior when using DatasetCreator with respect to sharding

### ISSUE THREAD

**Issue by bli00** [tensorflow/tensorflow#56228]
### Issue Type

Documentation Bug

### Source

binary

### Tensorflow Version

2.8.0

### Custom Code

No

### OS Platform and Distribution

Google Colab, tfx training using `gcr.io/tfx-oss-public/tfx:1.7.1`

### Current Behaviour?

I have an autoencoder training under the default `model.fit` using a `DatasetCreator`. My setup is 8 workers, 4 accelerator each, 8 ps.

When I add sharding to the `dataset_fn` in `DatasetCreator`, the loss MSE changes dramatically. Namely, calling (see code below):

```python
train_dataset = tf.keras.utils.experimental.DatasetCreator(
		lambda input_context: get_dataset(
				fn_args.train_files,
				transformed_feature_spec,
				preprocessor,
				batch_size=256,
				# remove sharding dramatically changes the loss mse
				shard_num=None,
				shard_index=None))
```

Gives me loss in the thousands (see logs), while calling:

```python
train_dataset = tf.keras.utils.experimental.DatasetCreator(
		lambda input_context: get_dataset(
				fn_args.train_files,
				transformed_feature_spec,
				preprocessor,
				batch_size=256,
				# remove sharding dramatically changes the loss mse
				shard_num=input_context.num_input_pipelines,
				shard_index=input_context.input_pipeline_id))
```

Gives me normal MSE less than 1 (see logs).

To debug, I added `tf.print()` to `get_dataset`, the printed logs values printed by the workers are:

```
num_input_pipelines == 1
input_pipeline_id == 0
num_replicas_in_sync == 4
```

Which tells me the `InputContext` given to each worker are unaware of other workers. And this makes sense in the context of the `ParameterServerStrategy` [tutorial guide][1], which states:

>Keras Model.fit with parameter server training assumes that each worker receives the same dataset, except when it is shuffled differently.

So I'm not sure how sharding the dataset with `dataset.shard(1, 0)` could possibly make such a difference. My only guess is that perhaps the debug `tf.print()` was triggered during tracing and the actual `InputContext` is different when training. **But even then, whether the data is sharded or not shouldn't affect the loss?** There's some unexpected behaviors here that's not documented?

[1]: https://www.tensorflow.org/tutorials/distribute/parameter_server_training#input_data

### Standalone code to reproduce the issue

```python
def get_dataset(
		file_pattern: List[Text],
		feature_spec: Dict[str, common_types.FeatureSpecType],
		preprocessing_layer: keras.Model,
		batch_size: int = 256,
		shard_num: int = None,
		shard_index: int = None) -> tf.data.Dataset:

	file_pattern = tf.convert_to_tensor(file_pattern)

	compression_type = dataset_util.detect_compression_type(file_pattern)

	dataset = tf.data.Dataset.list_files(
		file_pattern,
		shuffle=False)

	options = tf.data.Options()
	options.deterministic = False
	options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.OFF
	dataset = dataset.with_options(options)
	
	dataset = dataset.interleave(
		lambda filename: tf.data.TFRecordDataset(
			filename, 
			compression_type=compression_type),
		cycle_length=tf.data.AUTOTUNE,
		num_parallel_calls=tf.data.AUTOTUNE,
		deterministic=False)

	def _apply_preprocessing(tfrecord):
		parsed_features = tf.io.parse_example(tfrecord, feature_spec)
		preprocessed_features = tf.squeeze(preprocessing_layer(parsed_features))
		return preprocessed_features, preprocessed_features

	dataset = dataset.map(
		_apply_preprocessing,
		num_parallel_calls=tf.data.AUTOTUNE,
		deterministic=False)

	dataset = dataset.cache()

	dataset = dataset.shuffle(buffer_size=1000000)

	dataset = dataset.repeat()

	if shard_num and shard_index:
		dataset = dataset.shard(shard_num, shard_index)
	
	dataset = dataset.batch(
		batch_size,
		num_parallel_calls=tf.data.AUTOTUNE,
		deterministic=False)

	return dataset.prefetch(tf.data.AUTOTUNE)



# using a tfx pipeline
# TF_CONFIG is auto provisioned to the cluster
# 8 worker, 4 accelerator each, 8 ps, no eval workers
cluster_resolver = tf.distribute.cluster_resolver.TFConfigClusterResolver()

if cluster_resolver.task_type == 'chief':
	transformed_feature_spec = tf_transform_output.transformed_feature_spec()

	variable_partitioner = tf.distribute.experimental.partitioners.FixedShardsPartitioner(
			num_shards=8)

	strategy = tf.distribute.ParameterServerStrategy(
			cluster_resolver=cluster_resolver,
			variable_partitioner=variable_partitioner)


	with strategy.scope():
		preprocessor, autoencoder = create_model(
				feature_spec=transformed_feature_spec,
				learning_rate=0.004,
				)

		train_dataset = tf.keras.utils.experimental.DatasetCreator(
				lambda input_context: get_dataset(
						fn_args.train_files,
						transformed_feature_spec,
						preprocessor,
						batch_size=256,
						##############################################
						# Adding sharding dramatically reduce loss!? #
						##############################################
						shard_num=input_context.num_input_pipelines,
						shard_index=input_context.input_pipeline_id))

	autoencoder.fit(
			train_dataset,
			epochs=1024,
			steps_per_epoch=200)

	# save model...

elif cluster_resolver.task_type in ('worker', 'ps'):

	server = tf.distribute.Server(
			cluster_resolver.cluster_spec(),
			job_name=cluster_resolver.task_type,
			task_index=cluster_resolver.task_id,
			protocol='grpc',
			start=True)
	server.join()
```

### Relevant log output

---Without sharding:

```
workerpool0-0
Epoch 32/1024

workerpool0-0
200/200 - 4s - loss: 12659.2314 - mse: 12659.2275 - mae: 9.0156 - msle: 0.0395 - 4s/epoch - 19ms/step
```

---With sharding:

```
workerpool0-0
Epoch 32/1024

workerpool0-0
200/200 - 4s - loss: 0.3014 - mse: 0.2985 - mae: 0.3113 - msle: 0.0571 - 4s/epoch - 20ms/step
```

**Comment by bli00** [comment]
@mohantym any ideas here? I'm guessing it's just something that's not documented here as opposed to a behavior that's not actually not expected?

**Comment by sushreebarsa** [comment]
@bli00 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thank you!

**Comment by aljungberg** [comment]
Just an idea but I don't think this looks right:

```
if shard_num and shard_index:
    dataset = dataset.shard(shard_num, shard_index)
```

One of the shard indexes will be 0, so it won't use any sharding for that one generator. As a result, some of your training data is lost because the unsharded generator outputs the beginning of the sequence, rather than a strided subset.

**Comment by EEEGUI** [comment]
@feelingsonice  Have you fix this error ?
